### PR TITLE
251 issue/Provide_better_access_to_error_messages

### LINKED ISSUE #251: Provide better access to error messages
### Motivation

Provide better access to error messages, e.g. when an item is [not found](https://deliver.kontent.ai/975bf280-fd91-488c-994c-2f04416e5ee3/items/abcdefg), the user should have access to a strongly-typed error object to be able to differentiate the logic based on error codes.

And use e.g. the `catch-when` pattern: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/when

### Proposed solution

Enrich the [`IApiResponse`](https://github.com/Kentico/kontent-delivery-sdk-net/blob/master/Kentico.Kontent.Delivery.Abstractions/SharedModels/IApiResponse.cs) with an Error object containing the message, request id, error code, and specific code.

### More context
- [related intercom convo](https://app.intercom.com/a/apps/e42kus8l/inbox/inbox/conversation/19590600083373)

### PR DESCRIPTION
**Pull request by jirikk**
### Motivation

Which issue does this fix? Fixes #251

 - IApiResponse enriched by IError and IsSuccess properties 
 - Updated and extended unit tests
 -  small changes

### Checklist

- [x] Code follows coding conventions held in this repo
- [x] Automated tests have been added
- [x] Tests are passing
- [ ] Docs have been updated (if applicable)
- [] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

1. create a sample app
2. call DeliveryClient.GetItemAsync

### CODE REVIEW

File: Kentico.Kontent.Delivery/DeliveryClient.cs [hunk 1]
```diff
@@ -308,15 +314,24 @@ private async Task<ApiResponse> GetResponseContentAsync(HttpResponseMessage http
...
             {
-                faultContent = await httpResponseMessage.Content.ReadAsStringAsync();
+                var faultContent = await httpResponseMessage.Content.ReadAsStringAsync();
+                error = JsonConvert.DeserializeObject<Error>(faultContent);
+            }
+            else
+            {
+                error = new Error
+                {
+                    Message = $"Unknown error, HTTP statuc code: {httpResponseMessage.StatusCode}",
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery/DeliveryClient.cs:329]: "status"

File: Kentico.Kontent.Delivery.Abstractions/SharedModels/IApiResponse.cs [hunk 2]
```diff
@@ -26,5 +26,15 @@ public interface IApiResponse
         /// Gets the URL used to retrieve this response for debugging purposes.
         /// </summary>
         string RequestUrl { get; }
+
+        /// <summary>
+        /// Indicates whether call to Kentico Kontent Delivery API was successful or not.
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Abstractions/SharedModels/IApiResponse.cs:31]: "a call"

File: Kentico.Kontent.Delivery/DeliveryClient.cs [hunk 3]
```diff
@@ -308,15 +314,24 @@ private async Task<ApiResponse> GetResponseContentAsync(HttpResponseMessage http
...
 
-            string faultContent = null;
+            Error error;
 
             // The null-coalescing operator causes tests to fail for NREs, hence the "if" statement.
             if (httpResponseMessage?.Content != null)
             {
-                faultContent = await httpResponseMessage.Content.ReadAsStringAsync();
+                var faultContent = await httpResponseMessage.Content.ReadAsStringAsync();
+                error = JsonConvert.DeserializeObject<Error>(faultContent);
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery/DeliveryClient.cs:323]: use the instance of `JsonSerializer` (called Serializer) instead...

we try not to rely on the static classes

File: Kentico.Kontent.Delivery/DeliveryClient.cs [hunk 4]
```diff
@@ -368,7 +383,12 @@ private static void ValidateItemsFeedParameters(IEnumerable<IQueryParameter> par
 
         private async Task<IList<object>> GetLinkedItemsAsync(JObject content)
         {
-            var items = ((JObject)content["modular_content"]).Values().Select(async source => await ModelProvider.GetContentItemModelAsync<object>(source, content["modular_content"]));
+            var items = ((JObject)content["modular_content"])?.Values()?.Select(async source => await ModelProvider.GetContentItemModelAsync<object>(source, content?["modular_content"]));
+
+            if(items == null)
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery/DeliveryClient.cs:388]: why not inline these in the return statement? :)

File: Kentico.Kontent.Delivery.Abstractions/SharedModels/IError.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    /// </summary>
+    public interface IError
+    {
+        /// <summary>
+        /// Gets error Message.
+        /// </summary>
+        string Message { get; }
+
+        /// <summary>
+        /// Gets request id.
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Abstractions/SharedModels/IError.cs:14]: how about something like "Gets the ID of a request that can be used for troubleshooting"?

File: Kentico.Kontent.Delivery.Abstractions/SharedModels/IError.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        /// </summary>
+        string RequestId { get; }
+
+        /// <summary>
+        /// Gets error code.
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Abstractions/SharedModels/IError.cs:19]: I'd be a bit more specific..what kind of error code? HTTP status? KK-specific error code?

File: Kentico.Kontent.Delivery.Abstractions/SharedModels/IError.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
+﻿namespace Kentico.Kontent.Delivery.Abstractions
```
Comment (petrsvihlik) [inline Kentico.Kontent.Delivery.Abstractions/SharedModels/IError.cs:1]: generally, try to make the comments a bit more useful (and add the right articles the/a) :)
this has a crucial impact on the developer's experience...when they run into errors they'll appreciate any help narrowing down the root cause.

**Review by petrsvihlik — CHANGES_REQUESTED** [review]
just a few cosmetic changes

File: Kentico.Kontent.Delivery/DeliveryClient.cs (hunk 1 above)
Comment (jirikk) [reply Kentico.Kontent.Delivery/DeliveryClient.cs]: Removed

File: Kentico.Kontent.Delivery.Abstractions/SharedModels/IApiResponse.cs (hunk 2 above)
Comment (jirikk) [reply Kentico.Kontent.Delivery.Abstractions/SharedModels/IApiResponse.cs]: Fixed

File: Kentico.Kontent.Delivery/DeliveryClient.cs (hunk 3 above)
Comment (jirikk) [reply Kentico.Kontent.Delivery/DeliveryClient.cs]: Serializer used.

File: Kentico.Kontent.Delivery/DeliveryClient.cs (hunk 4 above)
Comment (jirikk) [reply Kentico.Kontent.Delivery/DeliveryClient.cs]: Fixed

File: Kentico.Kontent.Delivery.Abstractions/SharedModels/IError.cs (hunk 5 above)
Comment (jirikk) [reply Kentico.Kontent.Delivery.Abstractions/SharedModels/IError.cs]: Thank you for the suggestion, updated.

File: Kentico.Kontent.Delivery.Abstractions/SharedModels/IError.cs (hunk 6 above)
Comment (jirikk) [reply Kentico.Kontent.Delivery.Abstractions/SharedModels/IError.cs]: Could you please have a look now?

File: Kentico.Kontent.Delivery.Abstractions/SharedModels/IError.cs (hunk 7 above)
Comment (jirikk) [reply Kentico.Kontent.Delivery.Abstractions/SharedModels/IError.cs:1]: Noted

**Review by petrsvihlik — APPROVED** [review]

**Pull request merged**